docs/rules/no-mutation-props.md

```diff
@@ -0,0 +1,115 @@
+# Prevent mutation of this.props (no-mutation-props)
+
+NEVER mutate `this.props`, as all React components must act like pure functions with respect to their props. 
+Treat `this.props` as if it were immutable. More info available at [props-are-read-only](https://facebook.github.io/react/docs/components-and-props.html#props-are-read-only)
+
+## Rule Details
+
+This rule is aimed to forbid the use of mutating `this.props`.
+
+The following patterns are considered warnings:
+
+```jsx
+var Hello = React.createClass({
+  render: function() {
+    this.props.name = this.props.name.toUpperCase();
+    return <div>Hello {this.props.name}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    const {list} = this.props;
+    list.push(2);
+    return <div>{list.length} things</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    const [firstThing] = this.props.list;
+    firstThing.foo = 'bar';
+    return <div>{firstThing.foo}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    delete this.props.foo;
+    return <div>{Object.keys(this.props).length} props</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    Object.assign(this.props.foo, {bar: 'baz'});
+    return <div>{this.props.foo.bar}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    Object.defineProperty(this.props, 'foo');
+    return <div>{this.props.foo}</div>;
+  }
+});
+
+var Hello = React.createClass({
+  render: function() {
+    this.props.foo++;
+    return <div>{this.props.foo}</div>;
+  }
+});
+```
+
+## Rule Options
+
+```js
+...
+"react/no-mutation-props": [<enabled>, {
+  "allowArrayMutations": <boolean>,
+  "allowableArrayMutations": <array<string>>,
+  "allowObjectStatics": <boolean>,
+  "allowableObjectStatics": <array<string>>,
+}]
+...
+```
+
+### `allowArrayMutations`
+
+When `true` the rule ignores [array mutation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#Mutator_methods) of props.
```

Comment:
Boolean options should always default to `false`, so that absence and `false` are equivalent - this implies that anyone who hasn't passed an explicit `true` here, will get the warnings.

Regardless of what you want the default behavior to be, the booleans should be named such that they can default to `false`.

Separately, why do you think this behavior should be on by default, especially since there's caveats?

Reply:
Agreed with the first part. This does default to `false`, I can specify that if you think it needs to be clearer.

I like my tools to work out of the box. I really think that both of the cases you outlined are corner cases.

The chances of a `prop` that's an object, with methods, that also has methods named the same as array mutation methods are low. Frankly, doing so seems like bad practice (outside of immutable.js objects, but you shouldn't be calling the mutation methods in a component). 

Similarly, if you choose to override a language builtin like `Object.assign`, it's think it's reasonable for your linter to complain. As we've discussed, I prefer not to even have the object statics option – I think it encourages bad behavior.

The options are here to help you get around extremely unusual circumstances, and I struggle to see a case where they should ever be used.

Reply:
Defaults need to be safe; the instance methods in particular are not safe by default.

If the chance is nonzero, it's not an acceptable amount.

I'd be fine defaulting the static methods to be enabled, because as you say, the only way that'd be a problem is if someone intentionally broke the environment.

Reply:
Hmm… I'd argue it's more important for defaults to solve the common case, even if there's a little bit of danger involved. If I were using this rule, I wouldn't expect array mutations to not be reported.

I appreciate that point of view, and I'm not sure what your preferred method of resolving issues like this are. Maybe we can ask another maintainer to weigh in?

Reply:
Until other maintainers have weighed in, my vote remains that instance method mutations (which is a better name than "array method", and I'd like that changed also) are not safe to enable by default.

Anything less than 100% correct isn't a safe default.